### PR TITLE
gossiper: Graph announcements eviction

### PR DESCRIPTION
**Pull request by halseth**
This PR addresses a concern within the gossiper, where a malicious peer could serve us an excessive amounts of premature graph announcements, which we would store, leading to memory filling up. In case of an excessive amount of updates being stored, this is now addressed by

1.  evicting premature announcements that have a block height we don't know about yet, by removing those that have the greatest block height first. 
2. evicting `ChannelUpdates` that have the lowest timestamp. 

Both strategies involving evicting half of the stored updates if we reach the specified threshold.

### CODE REVIEW

File: discovery/gossiper.go [hunk 1]
```diff
@@ -928,16 +929,14 @@ func (d *AuthenticatedGossiper) networkHandler() {
 			// Next we check if we have any premature announcements
 			// for this height, if so, then we process them once
 			// more as normal announcements.
-			d.Lock()
 			numPremature := len(d.prematureAnnouncements[uint32(newBlock.Height)])
-			d.Unlock()
+			d.pAnnMtx.Unlock()
```
Comment (Roasbeef) [inline discovery/gossiper.go]: Looks like we're missing a `d.pAnnMtx.Unlock()` on the other side of this line. Same for the usage below.

Comment (halseth) [reply discovery/gossiper.go]: Hm, not sure what happened here... Fixed!

File: discovery/gossiper.go [hunk 2]
```diff
@@ -1226,17 +1233,146 @@ func (d *AuthenticatedGossiper) processRejectedEdge(chanAnnMsg *lnwire.ChannelAn
...
+			// Break out if we have deleted at least half of
+			// the announcments.
+			if d.numPrematureAnnouncements <= maxNumPrematureAnn/2 {
+				break
+			}
+		}
+	}
+
+	// Add the announcement to the map of premature announcements.
+	d.prematureAnnouncements[blockHeight] = append(
```
[the remark this answers is not in the document]
Comment (halseth) [reply discovery/gossiper.go]: We could do that... Rationale was that an incoming announcements would be potentially "newer" than what already found in the cache, such that we would evict "old" announcements. Not sure if it matters that much, this case should be uncommon (hopefully).

File: discovery/gossiper.go [hunk 3]
```diff
@@ -1226,17 +1233,146 @@ func (d *AuthenticatedGossiper) processRejectedEdge(chanAnnMsg *lnwire.ChannelAn
...
+		var updates []*networkMsg
+		for _, update := range d.prematureChannelUpdates {
+			updates = append(updates, update)
+		}
+
+		// Sort the updates, making the newest ones go last.
+		sort.Slice(updates, func(i, j int) bool {
+			upd1, ok := updates[i].msg.(*lnwire.ChannelUpdate)
+			if !ok {
+				return true
```
[the remark this answers is not in the document]
Comment (halseth) [reply discovery/gossiper.go]: To make them get evicted first.

**Comment by halseth** [comment]
> What advantage does evicting half of the cache have over using an eviction policy that operates iteratively?
This was done to make it harder for an attacker to spam us with announcements that would eat CPU time. Evicting half makes sure that he will need to feed us `O(N)` fake announcements before we do an evicting (costing us `O(n logn)` because of the sorting). Iterative eviction would be costlier I think.

File: discovery/gossiper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply discovery/gossiper.go]: Ah that makes sense :)

**Comment by Roasbeef** [comment]
Still relevant?

**Comment by halseth** [comment]
Still relevant, but the code is most likely very outdated now..

**Comment by Roasbeef** [comment]
Obsolete now, we no longer store premature announcement at all.

**Pull request closed without merging**